### PR TITLE
feat: switch svg

### PR DESCRIPTION
**Pull request by Jo-Chris**
`npx hardhat test`

prints out generated SVG ready to be viewed in browser.

Open points:

- [x] bar is always set to 100% 
- [x] [NFT](https://sepolia.etherscan.io/address/0xeef925ec7221ceb268a4b1050b6e3d238d618171)
- [x] [SvgGenerator](https://sepolia.etherscan.io/address/0xcced2ebefb3335ba830ca24fa912abbb7f7517a4#readContract)
- [x] randomness

image

Here's an example:

Bildschirmfoto 2024-06-22 um 15 25 46

### CODE REVIEW

**Comment by wsdt** [comment]
@Jo-Chris the bar on the bottom should be at 100%

File: contracts/nfts/SVGGenerator.sol [hunk 1]
```diff
@@ -76,71 +76,111 @@ contract SVGGenerator {
         "<line class='cls-11' x1='",rowMidpoint.toString(),"' y1='",TABLE_INITIAL_Y.toString(),"' x2='",rowMidpoint.toString(),"' y2='",tableEndY.toString(),"'/>"));
     }
 
+
+    function calculateCollateralLockedWidth(uint256 value) public view returns (string memory) {
```
Comment (wsdt) [inline contracts/nfts/SVGGenerator.sol:80]: I guess this can be private?

**Comment by Jo-Chris** [comment]
I've removed the basic color for the two rectancles inside, leading to this glass effect. Furthermore the svg size is a bit smaller now. 

These are the first 6 token ids. 

- [ ] The circle bottom left is still static. Is that something we should keep, we can also negate the background to get an contrast here.

image

**Review by ewansheldon — CHANGES_REQUESTED** [review]
really nice! i don't think there is a "nice" "readable" way to write these SVG generating contracts, but it is definitely better. the design is also looking much nicer

File: scripts/deployNft.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    let svgGenerator = await ethers.getContractFactory('SVGGenerator', {
+        libraries: {
+            NFTUtils: lib.address,
+        },
+    });
+
+    svgGenerator = await svgGenerator.deploy();
+    await svgGenerator.deployed();
+    console.log('SVGGenerator deployed to:', svgGenerator.address);
```
[the remark this answers is not in the document]
Comment (Jo-Chris) [reply scripts/deployNft.js]: now deploying the NftMetadatagenerator with [6f5f6a6](https://github.com/the-standard/smart-vault/pull/1/commits/6f5f6a6d6442b7f11f07f5bc8a496866ba9804fb)

File: test/svgGenerator.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        await svgGenerator.deployed();
+    });
+
+    it('should create an svg', async function () {
+        // Create a sample Status struct
+        const sampleStatus = {
+            vaultAddress: "0x1234567890123456789012345678901234567890",
+            minted: ethers.BigNumber.from("1000"),
+            maxMintable: ethers.BigNumber.from("10000"),
+            totalCollateralValue: ethers.BigNumber.from("500000"),
```
[the remark this answers is not in the document]
Comment (Jo-Chris) [reply test/svgGenerator.js]: adjusted the values in: [9f9d838](https://github.com/the-standard/smart-vault/pull/1/commits/9f9d838753722189ad013a62c2238ae7f1bad771)

File: contracts/nfts/SVGGenerator.sol [hunk 4]
```diff
@@ -76,71 +76,63 @@ contract SVGGenerator {
...
+                "<text x='325' y='528' font-weight='bold' text-anchor='middle'>&#8364; ", NFTUtils.toDecimalString(_vaultStatus.minted, 18), "</text>",
+                "<text x='470' y='490' text-anchor='middle'>Collateral/Debt</text>", "<rect x='419' y='504' width='132' height='40' rx='11' fill='#979DFA'/>",
+                "<text x='485' y='528' font-weight='bold' text-anchor='middle'>", collateralDebtPecentage(_vaultStatus), "</text>", "<text x='720' y='490' text-anchor='middle'>Total Minus Debt</text>",
+                "<rect x='662' y='504' width='132' height='40' rx='11' fill='url(#paint4_linear_428_47)'/>",
+                "<text x='730' y='528' font-weight='bold' text-anchor='middle'>&#8364;", NFTUtils.toDecimalString(_vaultStatus.totalCollateralValue - _vaultStatus.minted, 18), "</text>",
+                "<text x='221' y='622' font-size='18' text-anchor='middle'>Collateral locked in this vault</text>",
+                "<text x='790' y='628' font-size='18' font-weight='bold' text-anchor='end'>&#8364; N/V </text>", "<rect x='107' y='640' width='687' height='16' rx='8' fill='#AC99F7'/>",
```
[the remark this answers is not in the document]
Comment (Jo-Chris) [reply contracts/nfts/SVGGenerator.sol]: adjusted in [d7494a3](https://github.com/the-standard/smart-vault/pull/1/commits/d7494a3a2d7a5ae03c937d0a78796fc8ea6f0a4e)

Removing the comment will print the SVG in the test again to view it.

File: test/svgGenerator.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
+const {ethers} = require("hardhat");
+const {expect} = require('chai');
+
+const getSvgMintConfog = (minted, maxMintable, totalCollateralValue) => {
```
Comment (ewansheldon) [inline test/svgGenerator.js:4]: can we fix this typo

File: test/svgGenerator.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        maxMintable: ethers.BigNumber.from(ethers.utils.parseUnits(maxMintable, 18),),
+        totalCollateralValue: ethers.BigNumber.from(ethers.utils.parseUnits(totalCollateralValue, 18)),
+        collateral: [],
+        liquidated: false,
+        version: 1,
+        vaultType: ethers.utils.formatBytes32String("sampleVault")
+    }
+};
+
+describe.only('SVG Generator', async () => {
```
Comment (ewansheldon) [inline test/svgGenerator.js:17]: let's remove the test focus now this is ready to go

File: test/svgGenerator.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
+const {ethers} = require("hardhat");
+const {expect} = require('chai');
+
+const getSvgMintConfog = (minted, maxMintable, totalCollateralValue) => {
+    return {
+        vaultAddress: "0x1234567890123456789012345678901234567890",
+        minted: ethers.BigNumber.from(ethers.utils.parseUnits(minted, 18)),
+        maxMintable: ethers.BigNumber.from(ethers.utils.parseUnits(maxMintable, 18),),
+        totalCollateralValue: ethers.BigNumber.from(ethers.utils.parseUnits(totalCollateralValue, 18)),
```
Comment (ewansheldon) [inline test/svgGenerator.js:9]: i think `ethers.BigNumber.from` is superfluous here as `parseUnits` returns a big number. you can remove

**Review by ewansheldon — CHANGES_REQUESTED** [review]
last small changes but i think we're done 🎉

File: test/svgGenerator.js (hunk 7 above)
Comment (Jo-Chris) [reply test/svgGenerator.js]: referenced in: [8952a3b](https://github.com/the-standard/smart-vault/pull/1/commits/8952a3bb6e8a561e3b59a080b27a15986337f884)

File: test/svgGenerator.js (hunk 6 above)
Comment (Jo-Chris) [reply test/svgGenerator.js]: referenced in: [8952a3b](https://github.com/the-standard/smart-vault/pull/1/commits/8952a3bb6e8a561e3b59a080b27a15986337f884)

**Pull request merged**